### PR TITLE
Cluster Template Enforcement

### PR DESCRIPTION
**Pull request by prachidamle**
This PR will change after https://github.com/rancher/rancher/pull/20556 is merged.

### CODE REVIEW

File: pkg/api/customization/cluster/validator.go [hunk 1]
```diff
@@ -49,3 +70,80 @@ func (v *Validator) validateLocalClusterAuthEndpoint(request *types.APIContext,
...
+		return httperror.NewFieldAPIError(httperror.MissingRequired, "", "A clusterTemplate is required to create a cluster")
+	}
+
+	enforced, err := v.isTemplateEnforced(spec)
+	if err != nil {
+		return err
+	}
+
+	if !enforced {
+		return httperror.NewFieldAPIError(httperror.PermissionDenied, "", "This is not a valid clusterTemplate to create a cluster")
```
Comment (alena1108) [inline pkg/api/customization/cluster/validator.go]: nit: we should change an error message to include the template name.

File: pkg/api/customization/cluster/validator.go [hunk 2]
```diff
@@ -49,3 +70,80 @@ func (v *Validator) validateLocalClusterAuthEndpoint(request *types.APIContext,
...
+
+	clusterTemp, err := v.ClusterTemplateLister.Get(namespace.GlobalNamespace, spec.ClusterTemplateName)
+	if err != nil {
+		return false, err
+	}
+	if !clusterTemp.Spec.Enforced {
+		return false, nil
+	}
+
+	//make sure it is public
```
Comment (alena1108) [inline pkg/api/customization/cluster/validator.go]: can we just check the member attribute on the template? Would save us on the API call :)

File: pkg/api/customization/cluster/validator.go [hunk 3]
```diff
@@ -49,3 +68,93 @@ func (v *Validator) validateLocalClusterAuthEndpoint(request *types.APIContext,
...
+
+	isAdmin, err := ma.IsAdmin(callerID)
+	if err != nil {
+		return err
+	}
+	if isAdmin {
+		return nil
+	}
+
+	if !strings.EqualFold(settings.ClusterTemplateEnforcement.Get(), "true") {
```
Comment (alena1108) [inline pkg/api/customization/cluster/validator.go]: do you think we can make this check before the caller id? So we save on IsAdmin call :)

File: pkg/api/customization/cluster/validator.go [hunk 4]
```diff
@@ -49,3 +68,93 @@ func (v *Validator) validateLocalClusterAuthEndpoint(request *types.APIContext,
...
+		return nil
+	}
+	//enforcement is true, template is a must
+	if spec.ClusterTemplateID == "" || spec.ClusterTemplateRevisionID == "" {
```
Comment (alena1108) [inline pkg/api/customization/cluster/validator.go]: is this check already covered by checkClusterForEnforcement?

Comment (prachidamle) [reply pkg/api/customization/cluster/validator.go]: no, we need to make sure template is passed in case it is a create cluster call passing only rkeconfig

**Comment by alena1108** [comment]
LGTM

**Pull request merged**